x/wasm-storage/types/filters.go

```diff
@@ -0,0 +1,299 @@
+	case 0x03: // Uint64
+		return detectOutliersInteger[uint64](dataList, f.maxSigma)
+	default:
+		return nil, ErrInvalidNumberType
+	}
+}
+
+func detectOutliersInteger[T constraints.Integer](dataList []string, maxSigma uint64) ([]int, error) {
+	var nums []T
+	var corruptCount int
+	var corruptInds []int
+	var z T
+	rt := reflect.TypeOf(z)
+	length := len(dataList)
+	for i, data := range dataList {
+		if data == "" {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+
+		bz, err := base64.StdEncoding.DecodeString(data)
+		if err != nil {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+		if len(bz) != int(rt.Size()) {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+
+		switch rt.Kind() {
+		case reflect.Int32:
+			var num int32
+			err = binary.Read(bytes.NewBuffer(bz), binary.BigEndian, &num)
+			if err != nil {
+				corruptCount++
+				corruptInds = append(corruptInds, i)
+				continue
+			}
+			nums = append(nums, T(num))
+
+		case reflect.Int64:
+			var num int64
+			err = binary.Read(bytes.NewBuffer(bz), binary.BigEndian, &num)
+			if err != nil {
+				corruptCount++
+				corruptInds = append(corruptInds, i)
+				continue
+			}
+			nums = append(nums, T(num))
+
+		case reflect.Uint32:
+			num := binary.BigEndian.Uint32(bz)
+			nums = append(nums, T(num))
+
+		case reflect.Uint64:
+			num := binary.BigEndian.Uint64(bz)
+			nums = append(nums, T(num))
+
+		default:
+			panic("invalid number type") // TODO should never end up here
+		}
+	}
+
+	// If more than 1/3 of the reveals are corrupted,
+	// we reach consensus that the reveals are unusable.
+	if corruptCount*3 > length {
+		return nil, ErrCorruptReveals
+	}
+
+	// Sort and find median.
+	numsSorted := make([]T, len(nums))
+	copy(numsSorted, nums)
+	slices.Sort(numsSorted)
+
+	var median T
+	var medianHalf bool // if true, median must be corrected by adding 0.5
+	if length%2 == 1 {
+		median = numsSorted[length/2]
+	} else {
+		mid := length / 2
+		median = (numsSorted[mid-1] + numsSorted[mid]) / 2
+		if (numsSorted[mid-1]%2 == 0 && numsSorted[mid]%2 == 1) ||
+			(numsSorted[mid-1]%2 == 1 && numsSorted[mid]%2 == 0) {
+			medianHalf = true
+		}
+	}
+
+	// Identify outliers and keep their count.
+	outliers := make([]int, len(dataList))
+	var nonOutlierCount int
+	i, j := 0, 0 // i tracks nums and j tracks corruptInds.
+	for i+j < len(dataList) {
+		if corruptCount > 0 && i+j == corruptInds[j] {
+			outliers[i+j] = 1
+			corruptCount--
+			j++
+			continue
+		}
+		if isOutlier(maxSigma, nums[i], median, medianHalf) {
+			outliers[i+j] = 1
+		} else {
+			nonOutlierCount++
+		}
+		i++
+	}
+
+	// If less than 2/3 of the numbers fall within max sigma range
+	// from the median, there is no consensus.
+	if nonOutlierCount*3 < len(nums)*2 {
+		return outliers, ErrNoConsensus
+	}
+	return outliers, nil
+}
+
+func isOutlier[T constraints.Integer](maxSigma uint64, num, median T, medianHalf bool) bool {
```

Comment:
I think we should have one function that calculates the Sigma (in a seperate file) that doesn't know about reveals or protocol logic (apart from the 6 decimals). And together with the medianHalf logic. Right now its a lil splinters across functions while the logic is one thing as a whole.

Reply:
I feel bad for making you review this now 😬😬 I added `Sigma` type in [bb15b42](https://github.com/sedaprotocol/seda-chain/pull/300/commits/bb15b42eedf6d60aebc92a204ce84a7db336450b) and `HalfStepInt` in [22a571d](https://github.com/sedaprotocol/seda-chain/pull/300/commits/22a571d2a4c7ed40e0ec2fc04e7898f62c6c0c40)